### PR TITLE
[GNA] Improve scores exporting

### PR DESCRIPTION
**Pull request by mryzhov**
### Details:
Export loops have been fused to improve performance
Expensive operations have been moved out of loops

### Tickets:
 - *78822*

### CODE REVIEW

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 1]
```diff
@@ -210,69 +234,44 @@ void GNAPlugin::ExportScores(void *ptr_dst,
...
-                    case Precision::I32 : {
-                        *dst_ptr = *reinterpret_cast<const int32_t *>(input_ptr);
-                        break;
-                    }
-                    default:
-                        THROW_GNA_EXCEPTION << "Unsupported output layer precision: " << precision_in.name();
-                }
+                auto dst_ptr = ptr_dst + (i * num_vector_elements + j);
+                auto input_ptr = src + (j * num_group + i) * precision_in_size;
+                UpscaleAndCast(dst_ptr, input_ptr, precision_in, scale_factor);
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:258]: IMO calling `UpscaleAndCast` in a loop may still lead to poor accuracy ...
The compiler may still use branches for switch/case on `precision_in` for every converted number.

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 2]
```diff
@@ -201,7 +201,31 @@ void GNAPlugin::copyInputData(T *dst,
...
+            break;
+        }
+        default:
+            THROW_GNA_EXCEPTION << "Unsupported output layer precision: " << precision_in.name();
+    }
+    if (gnadevice) {
+        *dst_ptr /= scale_factor;
+    }
+}
+
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:226]: IMO we should not complicate gna_plugin.cpp (already large file) with such numeric code.

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 3]
```diff
@@ -201,7 +201,31 @@ void GNAPlugin::copyInputData(T *dst,
     }
 }
 
-void GNAPlugin::ExportScores(void *ptr_dst,
+template <typename T>
+void GNAPlugin::UpscaleAndCast(T* dst_ptr, const uint8_t* input_ptr, const Precision& precision_in, const float& scale_factor) {
+    switch (precision_in) {
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:206]: we should avoid switch/is etc in this perf-cruitical function
we should move it outside a loop.

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 4]
```diff
@@ -201,7 +201,31 @@ void GNAPlugin::copyInputData(T *dst,
...
+            break;
+        }
+        case Precision::I32: {
+            *dst_ptr = static_cast<T>(*reinterpret_cast<const int32_t*>(input_ptr));
+            break;
+        }
+        default:
+            THROW_GNA_EXCEPTION << "Unsupported output layer precision: " << precision_in.name();
+    }
+    if (gnadevice) {
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:222]: we should avoid switch/is etc in this perf-cruitical function
we should move it outside a loop.

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 5]
```diff
@@ -201,7 +201,31 @@ void GNAPlugin::copyInputData(T *dst,
...
+        }
+        case Precision::I16: {
+            *dst_ptr = static_cast<T>(*reinterpret_cast<const int16_t*>(input_ptr));
+            break;
+        }
+        case Precision::I32: {
+            *dst_ptr = static_cast<T>(*reinterpret_cast<const int32_t*>(input_ptr));
+            break;
+        }
+        default:
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:219]: this exception should be pouitside of a looop

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 6]
```diff
@@ -201,7 +201,31 @@ void GNAPlugin::copyInputData(T *dst,
     }
 }
 
-void GNAPlugin::ExportScores(void *ptr_dst,
+template <typename T>
+void GNAPlugin::UpscaleAndCast(T* dst_ptr, const uint8_t* input_ptr, const Precision& precision_in, const float& scale_factor) {
```
Comment (kbruniec) [inline src/plugins/intel_gna/src/gna_plugin.cpp:205]: IMO U**n**scale was a better name ;)
Upscale suggest that the resulting number is greater,
but input is **divided** by scale_factor
typically scale_factor is greater than 1

`CastAndDivide` is also reasonable name

**Comment by nkogteva** [comment]
One more PR?! How about this one https://github.com/openvinotoolkit/openvino/pull/13599?

**Comment by mryzhov** [comment]
Closing as all improvements will be finished in https://github.com/openvinotoolkit/openvino/pull/13599

**Pull request closed without merging**